### PR TITLE
Speed from O(N) to O(1) in ~GrpcMuxWatchImpl() and update().

### LINKED ISSUE #15528: ~GrpcMuxWatchImpl() uses watches_.remove(this) which traverses the list.
*Title*: ~GrpcMuxWatchImpl() uses watches_.remove(this) which traverses the list.

*Description*:

~GrpcMuxWatchImpl() uses std::list::remove() to remove itself from `watches_` list. 
This traverses the list and takes O(N) for each destruction.
We could speed things up to O(1) if GrpcMuxWatchImpl would keep a list iterator for itself and used std::list::erase() instead.

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

Commit Message: Speed from O(N) to O(1) in ~GrpcMuxWatchImpl() and update().
Additional Description: Please see https://github.com/envoyproxy/envoy/issues/15528
Risk Level: 
Testing: Format and Unit
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features:
Fixes #15528

### CODE REVIEW

File: source/common/config/grpc_mux_impl.h [hunk 1]
```diff
@@ -87,24 +87,26 @@ class GrpcMuxImpl : public GrpcMux,
           parent_(parent), watches_(parent.apiStateFor(type_url).watches_) {
       std::copy(resources.begin(), resources.end(), std::inserter(resources_, resources_.begin()));
       watches_.emplace(watches_.begin(), this);
+      it_ = watches_.begin();
```
Comment (jmarantz) [inline source/common/config/grpc_mux_impl.h:90]:
```
iter_ = watches_.emplace(watches_.begin(), this);
```
`it_` seems too terse for a member var.

File: source/common/config/grpc_mux_impl.h [hunk 2]
```diff
@@ -117,6 +119,7 @@ class GrpcMuxImpl : public GrpcMux,
 
   private:
     std::list<GrpcMuxWatchImpl*>& watches_;
+    std::list<GrpcMuxWatchImpl*>::iterator it_;
```
Comment (jmarantz) [inline source/common/config/grpc_mux_impl.h:122]: suggest making a `using` nickname for the list rather than typing it twice, but not a big deal.

**Review by jmarantz — APPROVED** [review]
looks great; thanks.

File: source/common/config/grpc_mux_impl.h (hunk 1 above)
Comment (mk46) [reply source/common/config/grpc_mux_impl.h:90]: I will do that.

File: source/common/config/grpc_mux_impl.h (hunk 2 above)
Comment (mk46) [reply source/common/config/grpc_mux_impl.h:122]: Thanks for the input. Is something like that `using GrpcMuxWatchImplList = std::list<GrpcMuxWatchImpl*>;` ?

Comment (jmarantz) [reply source/common/config/grpc_mux_impl.h]: Yes though considering this is local to the class, I'd just call it `WatchList`.

Comment (mk46) [reply source/common/config/grpc_mux_impl.h]: Done.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**